### PR TITLE
122 new controller integration

### PR DESCRIPTION
**Pull request by gdoffe**
Next of PR #84 .
Integrate new controller inside applications/platforms

### CODE REVIEW

File: platforms/pegasus/include/motion_control.hpp [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+///  - pulse_per_mm = wheels_encoder_resolution / wheels_perimeter
+///
+///  - wheels_diameter = 60 mm
+///  - wheels_distance_mm = 280 mm
+///  - wheels_encoder_resolution = 2000
+///
+/// @{
+constexpr double robot_width = 330;                 ///< Robot width (mm)
+constexpr double robot_margin = (robot_width / 2);  ///< Point the most far from robot center (mm)
+constexpr double beacon_support_diameter = 70;      ///< Size of the beacon support (a cylinder of 70mm diameter to a cube of 100mm width)
```
Comment (ecourtois) [inline platforms/pegasus/include/motion_control.hpp]: Is it still used?

File: platforms/pegasus/include/motion_control.hpp [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+/// Print robot state
+void pf_print_state(void);
+
+/// Compute current speed and pose
+void compute_current_speed_and_pose(
+    cogip::cogip_defs::Polar &current_speed,    ///< [out]  robot current speed
+    cogip::cogip_defs::Pose &current_pose       ///< [out]  robot current pose
+    );
+
+
```
Comment (ecourtois) [inline platforms/pegasus/include/motion_control.hpp:118]: Extra empty line.

File: platforms/pegasus/include/motion_control.hpp [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+/// Apply the given command to the motors
+void pf_motor_drive(
+    const cogip::cogip_defs::Polar &command     ///< [in]   linear and angular speeds command
+    );
+
+} // namespace actuators
+
+} // namespace pf
+
+} // namespace cogip
```
Comment (ecourtois) [inline platforms/pegasus/include/motion_control.hpp]: Missing empty line at EOF.

File: platforms/pegasus/motion_control.cpp [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+                << "\"O\":" << pf_motion_control_platform_engine.current_pose().O()
+                << ",\"x\":" << pf_motion_control_platform_engine.current_pose().x()
+                << ",\"y\":" << pf_motion_control_platform_engine.current_pose().y()
+            << "},"
+            << "\"cycle\":" << pf_motion_control_platform_engine.current_cycle() << ","
+            << "\"speed_current\":{\"distance\":" << pf_motion_control_platform_engine.current_speed().distance()
+            << ",\"angle\":" << pf_motion_control_platform_engine.current_speed().angle() << "}"
+    );
+
+    COGIP_DEBUG_COUT("}");
```
Comment (ecourtois) [inline platforms/pegasus/motion_control.cpp:290]: This must be included in the previous statement, but I think this function can be removed.

File: platforms/pegasus/motion_control.cpp [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+    // Start engine thread
+    pf_motion_control_platform_engine.start_thread();
+}
+
+void pf_encoder_read(cogip::cogip_defs::Polar &current_speed)
+{
+    int32_t left_speed = qdec_read_and_reset(MOTOR_LEFT) * QDEC_LEFT_POLARITY;
+    int32_t right_speed = qdec_read_and_reset(MOTOR_RIGHT) * QDEC_RIGHT_POLARITY;
+
+    /* update speed */
```
Comment (ecourtois) [inline platforms/pegasus/motion_control.cpp:384]: C++ comments

File: platforms/pegasus/motion_control.cpp [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+        cogip::pf::motion_control::pid_request_uuid,
+        cogip::uartpb::message_handler_t::create<_handle_pid_request>()
+    );
+}
+
+} // namespace actuators
+
+} // namespace pf
+
+} // namespace cogip
```
Comment (ecourtois) [inline platforms/pegasus/motion_control.cpp]: Missing empty line at EOF.

File: platforms/pf_test/include/motion_control.hpp [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+constexpr cogip::uartpb::uuid_t start_pose_uuid = 2741980922;
+constexpr cogip::uartpb::uuid_t state_uuid = 3422642571;
+constexpr cogip::uartpb::uuid_t pid_uuid = 4159164681;
+constexpr cogip::uartpb::uuid_t pid_request_uuid = 3438831927;
+
+constexpr uint16_t motion_control_thread_period_ms = 20;
+
+/// @name Robot mechanical properties
+///
+/// To be computed :
```
Comment (ecourtois) [inline platforms/pf_test/include/motion_control.hpp:38]: Extra space before :

File: platforms/pf_test/include/motion_control.hpp [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+///  - pulse_per_mm             : number of pulses per mm of coding wheel
+///  - wheels_distance          : distance between coding wheels (pulses)
+///  - pulse_per_degree         : number of pulses per degree of coding wheel
+///
+/// must be known :
```
Comment (ecourtois) [inline platforms/pf_test/include/motion_control.hpp:43]: Extra space before :

File: platforms/pf_test/include/motion_control.hpp [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+/// Apply the given command to the motors
+void pf_motor_drive(
+    const cogip::cogip_defs::Polar &command     ///< [in]   linear and angular speeds command
+    );
+
+} // namespace actuators
+
+} // namespace pf
+
+} // namespace cogip
```
Comment (ecourtois) [inline platforms/pf_test/include/motion_control.hpp:128]: Missing empty line at EOF.

File: platforms/pf_test/include/motion_control.hpp [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+///  - pulse_per_mm = wheels_encoder_resolution / wheels_perimeter
+///
+///  - wheels_diameter = 60 mm
+///  - wheels_distance_mm = 280 mm
+///  - wheels_encoder_resolution = 2000
+///
+/// @{
+constexpr double robot_width = 333.54;              ///< Robot width (mm)
+constexpr double robot_margin = (robot_width / 2);  ///< Point the most far from robot center (mm)
+constexpr double beacon_support_diameter = 70;      ///< Size of the beacon support (a cylinder of 70mm diameter to a cube of 100mm width)
```
Comment (ecourtois) [inline platforms/pf_test/include/motion_control.hpp]: Still used?

File: platforms/pf_test/motion_control.cpp [hunk 11]
```diff
@@ -0,0 +1,343 @@
...
+);
+// Angular speed PID controller
+static cogip::pid::PID angular_speed_pid(
+    angular_speed_pid_kp,
+    angular_speed_pid_ki,
+    angular_speed_pid_kd,
+    angular_speed_pid_integral_limit
+    );
+
+
```
Comment (ecourtois) [inline platforms/pf_test/motion_control.cpp:83]: Extra empty line.

File: platforms/pf_test/motion_control.cpp [hunk 12]
```diff
@@ -0,0 +1,343 @@
...
+                << ",\"x\":" << pf_motion_control_platform_engine.current_pose().x()
+                << ",\"y\":" << pf_motion_control_platform_engine.current_pose().y()
+            << "},"
+            << "\"cycle\":" << pf_motion_control_platform_engine.current_cycle() << ","
+            << "\"speed_current\":{\"distance\":" << pf_motion_control_platform_engine.current_speed().distance()
+            << ",\"angle\":" << pf_motion_control_platform_engine.current_speed().angle() << "}"
+    );
+
+    COGIP_DEBUG_COUT("}");
+}
```
Comment (ecourtois) [inline platforms/pf_test/motion_control.cpp:291]: Same as pegasus.

File: platforms/pf_test/motion_control.cpp [hunk 13]
```diff
@@ -0,0 +1,343 @@
...
+        pid_request_uuid,
+        cogip::uartpb::message_handler_t::create<_handle_pid_request>()
+    );
+}
+
+} // namespace actuators
+
+} // namespace pf
+
+} // namespace cogip
```
Comment (ecourtois) [inline platforms/pf_test/motion_control.cpp]: Missing empty at EOF.

File: platforms/pf_test/platform.cpp [hunk 14]
```diff
@@ -25,10 +27,33 @@
...
+        gpio_set(HEARTBEAT_LED);
+        ztimer_sleep(ZTIMER_MSEC, 200);
+        gpio_clear(HEARTBEAT_LED);
+        ztimer_sleep(ZTIMER_MSEC, 200);
+        gpio_set(HEARTBEAT_LED);
+        ztimer_sleep(ZTIMER_MSEC, 200);
+        gpio_clear(HEARTBEAT_LED);
+
+        // Wait thread period to end
+        ztimer_periodic_wakeup(ZTIMER_MSEC, &loop_start_time, 1800);
```
Comment (ecourtois) [inline platforms/pf_test/platform.cpp:51]: No define for thread period?

File: platforms/pf_test/motion_control.cpp [hunk 15]
```diff
@@ -326,6 +328,41 @@ static void _handle_new_pids_config(cogip::uartpb::ReadBuffer & buffer)
...
+    EmbeddedProto::Error error = pb_controller.deserialize(buffer);
+    if (error != EmbeddedProto::Error::NO_ERRORS) {
+        std::cout << "Controller change request: Protobuf deserialization error: " << static_cast<int>(error) << std::endl;
+        return;
+    }
+
+    // Change controller
+    std::cout << "Change to controller " << static_cast<uint32_t>(pb_controller.id()) << std::endl;
+    switch (static_cast<uint32_t>(pb_controller.id()))
+    {
```
Comment (ecourtois) [inline platforms/pf_test/motion_control.cpp:489]: { at end of previous line.

**Review by ecourtois — CHANGES_REQUESTED** [review]
First review.
Some typos, comments and questions, but I think we're almost good.

Commits b928b41 and ead43d6 to fixup.
Commit c550bce no signed.

File: boards/cogip-board-ng/board.c [hunk 16]
```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright (C) 2018 COGIP Robotics association
```
[the remark this answers is not in the document]
Comment (gdoffe) [reply boards/cogip-board-ng/board.c:2]: done

File: boards/cogip-board-ng/include/board.h [hunk 17]
```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (C) 2018 COGIP Robotics association
```
[the remark this answers is not in the document]
Comment (gdoffe) [reply boards/cogip-board-ng/include/board.h:2]: done

File: motion_control/engines/platform_engine/include/platform_engine/PlatformEngine.hpp [hunk 18]
```diff
@@ -36,10 +36,41 @@ class PlatformEngine : public BaseControllerEngine {
...
+    /// return     target speed
+    cogip_defs::Polar target_speed() const { return target_speed_; };
+
+    /// Get current pose
+    /// return     current pose
+    cogip_defs::Pose current_pose() const { return current_pose_; };
+
+    /// Get target pose
+    /// return     target pose
+    cogip_defs::Pose target_pose() const { return target_pose_; };
```
[the remark this answers is not in the document]
Comment (gdoffe) [reply motion_control/engines/platform_engine/include/platform_engine/PlatformEngine.hpp:57]: good point ! done !

File: motion_control/engines/platform_engine/include/platform_engine/PlatformEngine.hpp [hunk 19]
```diff
@@ -25,16 +25,20 @@ namespace motion_control {
...
+/// Process motion control commands
+using platform_process_commands_cb_t = etl::delegate<void(const cogip_defs::Polar&)>;
 
 /// Engine getting inputs from platform and setting outputs for the platform
 class PlatformEngine : public BaseControllerEngine {
 public:
     /// Constructor
     PlatformEngine(
         platform_get_speed_and_pose_cb_t platform_get_speed_and_pose_cb,    ///< [in]  Platform callback to get robot current speed and pose
+        platform_process_commands_cb_t platform_process_commands_cb         ///< [in]  Platform callback to get robot current speed and pose
```
[the remark this answers is not in the document]
Comment (gdoffe) [reply motion_control/engines/platform_engine/include/platform_engine/PlatformEngine.hpp:37]: done

File: motion_control/engines/platform_engine/include/platform_engine/PlatformEngine.hpp [hunk 20]
```diff
@@ -36,10 +36,41 @@ class PlatformEngine : public BaseControllerEngine {
...
+    /// Set current pose
+    void set_current_pose(
+        cogip_defs::Pose current_pose        ///< [in]   new current pose
+        ) { current_pose_ = current_pose; };
+
+    /// Set target pose
+    void set_target_pose(
+        cogip_defs::Pose target_pose        ///< [in]   new target pose
+        ) { target_pose_ = target_pose; };
+
```
[the remark this answers is not in the document]
Comment (gdoffe) [reply motion_control/engines/platform_engine/include/platform_engine/PlatformEngine.hpp:77]: done

File: motion_control/motion_control_common/include/motion_control_common/BaseController.hpp [hunk 21]
```diff
@@ -22,6 +22,12 @@ namespace cogip {
 
 namespace motion_control {
 
+/// Status of a position to reach
+/// - reached:      the position has been reached
+/// - intermediate: a transient position has been reached
+/// - ongoing:      moving
+enum target_pose_status_t {moving = 0, reached, intermediate_reached};
```
[the remark this answers is not in the document]
Comment (gdoffe) [reply motion_control/motion_control_common/include/motion_control_common/BaseController.hpp:29]: Cause it cannot be casted as a double to be passed in the controller chain arguments.

[the remark this answers is not in the document]
Comment (gdoffe) [reply motion_control/motion_control_common/include/motion_control_common/BaseController.hpp:29]: Which prefix/suffix then ?

[the remark this answers is not in the document]
Comment (gdoffe) [reply motion_control/motion_control_common/include/motion_control_common/BaseController.hpp:29]: `typedef enum {moving = 0, reached, intermediate_reached} target_pose_status_t;`

File: lib/path/Path.cpp [hunk 22]
```diff
@@ -32,12 +32,12 @@ size_t Path::operator--(int)
 
 double Path::current_max_speed_linear()
 {
-    return std::min(at(current_pose_index_).max_speed_linear(), (double)MAX_SPEED_LINEAR);
+    return std::min(at(current_pose_index_).max_speed_linear(), (double)cogip::pf::motion_control::platform_max_speed_linear_mm_per_period);
```
[the remark this answers is not in the document]
Comment (gdoffe) [reply lib/path/Path.cpp:35]: done

File: platforms/pegasus/include/motion_control.hpp [hunk 23]
```diff
@@ -0,0 +1,131 @@
...
+constexpr cogip::uartpb::uuid_t start_pose_uuid = 2741980922;
+constexpr cogip::uartpb::uuid_t state_uuid = 3422642571;
+constexpr cogip::uartpb::uuid_t pid_uuid = 4159164681;
+constexpr cogip::uartpb::uuid_t pid_request_uuid = 3438831927;
+
+constexpr uint16_t motion_control_thread_period_ms = 20;
+
+/// @name Robot mechanical properties
+///
+/// To be computed :
```
[the remark this answers is not in the document]
Comment (gdoffe) [reply platforms/pegasus/include/motion_control.hpp:38]: done

File: platforms/pegasus/include/motion_control.hpp [hunk 24]
```diff
@@ -0,0 +1,131 @@
...
+///  - pulse_per_mm             : number of pulses per mm of coding wheel
+///  - wheels_distance          : distance between coding wheels (pulses)
+///  - pulse_per_degree         : number of pulses per degree of coding wheel
+///
+/// must be known :
```
[the remark this answers is not in the document]
Comment (gdoffe) [reply platforms/pegasus/include/motion_control.hpp:43]: done

File: platforms/pegasus/include/motion_control.hpp (hunk 1 above)
Comment (gdoffe) [reply platforms/pegasus/include/motion_control.hpp]: no has been removed in a further commit
so done

File: platforms/pegasus/include/motion_control.hpp (hunk 2 above)
Comment (gdoffe) [reply platforms/pegasus/include/motion_control.hpp:118]: done

File: platforms/pegasus/motion_control.cpp (hunk 4 above)
Comment (gdoffe) [reply platforms/pegasus/motion_control.cpp:290]: done

Comment (gdoffe) [reply platforms/pegasus/motion_control.cpp:290]: can and should be removed but in a next PR ;)

File: platforms/pegasus/motion_control.cpp (hunk 5 above)
Comment (gdoffe) [reply platforms/pegasus/motion_control.cpp:384]: done

File: platforms/pegasus/motion_control.cpp (hunk 6 above)
Comment (gdoffe) [reply platforms/pegasus/motion_control.cpp]: weird, got them but in a further commit maybe

File: platforms/pf_test/include/motion_control.hpp (hunk 7 above)
Comment (gdoffe) [reply platforms/pf_test/include/motion_control.hpp:38]: done

File: platforms/pf_test/include/motion_control.hpp (hunk 9 above)
Comment (gdoffe) [reply platforms/pf_test/include/motion_control.hpp:128]: same weird, got them but in a further commit maybe

File: platforms/pf_test/include/motion_control.hpp (hunk 10 above)
Comment (gdoffe) [reply platforms/pf_test/include/motion_control.hpp]: no removed in a further commit

File: platforms/pf_test/motion_control.cpp (hunk 11 above)
Comment (gdoffe) [reply platforms/pf_test/motion_control.cpp:83]: done

File: platforms/pf_test/motion_control.cpp (hunk 12 above)
Comment (gdoffe) [reply platforms/pf_test/motion_control.cpp:291]: done

File: platforms/pf_test/motion_control.cpp (hunk 13 above)
Comment (gdoffe) [reply platforms/pf_test/motion_control.cpp]: done in further commit

File: platforms/pf_test/motion_control.cpp (hunk 15 above)
Comment (gdoffe) [reply platforms/pf_test/motion_control.cpp:489]: done

File: platforms/pf_test/platform.cpp (hunk 14 above)
Comment (gdoffe) [reply platforms/pf_test/platform.cpp:51]: :fearful: done

File: boards/cogip-board-ng/include/board.h (hunk 17 above)
[the remark this answers is not in the document]
Comment (ecourtois) [reply boards/cogip-board-ng/include/board.h]: Nope...

File: boards/cogip-board-ng/board.c (hunk 16 above)
[the remark this answers is not in the document]
Comment (ecourtois) [reply boards/cogip-board-ng/board.c:2]: Nope...

File: platforms/pf_test/platform.cpp (hunk 14 above)
Comment (ecourtois) [reply platforms/pf_test/platform.cpp]: Yes, but you used `static uint16_t` instead of `constexpr`.

**Review by ecourtois — APPROVED** [review]
Still 2 conversations to handle and 2 commits to check (df9a773d7 and f544356b5) but I already approve.

File: boards/cogip-board-ng/board.c (hunk 16 above)
[the remark this answers is not in the document]
Comment (gdoffe) [reply boards/cogip-board-ng/board.c:2]: done

File: platforms/pf_test/platform.cpp (hunk 14 above)
Comment (gdoffe) [reply platforms/pf_test/platform.cpp]: done

**Pull request merged**